WordPress/Classes/Services/ThemeService.m

```diff
@@ -263,32 +162,42 @@ - (NSProgress *)getThemesForBlog:(Blog *)blog
-                                       for (Theme *theme in unsyncedThemes) {
-                                           if (theme.custom) {
-                                               [unsyncedWPThemes removeObject:theme];
+                                   NSArray * __block themeObjectIDs = nil;
+                                   [self.coreDataStack performAndSaveUsingBlock:^(NSManagedObjectContext *context) {
+                                       Blog *blogInContext = [context existingObjectWithID:blog.objectID error:nil];
+                                       NSArray *themes = [self themesFromRemoteThemes:remoteThemes
+                                                                               forBlog:blogInContext
+                                                                               inContext:context];
+                                       if (sync) {
+                                           NSMutableSet *unsyncedThemes = [NSMutableSet setWithSet:blogInContext.themes];
+                                           // We don't want to touch custom themes here, only WP.com themes
+                                           NSMutableSet *unsyncedWPThemes = [unsyncedThemes mutableCopy];
+                                           for (Theme *theme in unsyncedThemes) {
+                                               if (theme.custom) {
+                                                   [unsyncedWPThemes removeObject:theme];
+                                               }
                                            }
-                                       }
-                                       [unsyncedWPThemes minusSet:[NSSet setWithArray:themes]];
-                                       for (Theme *deleteTheme in unsyncedWPThemes) {
-                                           if (![blog.currentThemeId isEqualToString:deleteTheme.themeId]) {
-                                               [self.managedObjectContext deleteObject:deleteTheme];
+                                           [unsyncedWPThemes minusSet:[NSSet setWithArray:themes]];
+                                           for (Theme *deleteTheme in unsyncedWPThemes) {
+                                               if (![blogInContext.currentThemeId isEqualToString:deleteTheme.themeId]) {
+                                                   [context deleteObject:deleteTheme];
+                                               }
                                            }
                                        }
-                                   }
-
-                                   [[ContextManager sharedInstance] saveContext:self.managedObjectContext withCompletionBlock:^{
+                                       [context obtainPermanentIDsForObjects:themes error:nil];
+                                       themeObjectIDs = [themes wp_map:^id(Theme *obj) {
```

Comment:
I don't think I ever noticed these custom functions before. They're neat, but I wonder if we'd be betters served in using a third party library for this.

Or rather, I wonder if we'd be better served moving all the code that needs map, filter, reduce, etc. to the Swift layer where they're already part of the standard library and we don't need custom or third-party code to use them. 😁 

_One day..._

Reply:
FWIW: https://github.com/wordpress-mobile/WordPress-iOS/issues/20185

Reply:
👍 for moving to Swift, which brings in whole lots of nice language features other than the array/collection operators.